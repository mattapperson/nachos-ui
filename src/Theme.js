import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
