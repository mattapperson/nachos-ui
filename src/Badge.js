import React from "react";
import PropTypes from "prop-types";
import { Text, View } from "react-native";
import { withTheme } from "./Theme";

const Badge = props => {
  const { color, style, textStyle, value, theme } = props;

  return (
    <View style={theme.container}>
      <View
        style={[theme.base, color ? { backgroundColor: color } : {}, style]}
      >
        <Text style={[theme.text, textStyle]}>{value}</Text>
      </View>
    </View>
  );
};

Badge.themeConfig = {
  style: {
    container: {
      flexDirection: "row"
    },
    base: {
      padding: 10,
      paddingVertical: 4,
      backgroundColor: "@accentColor",
      borderRadius: 20,
      minWidth: 40
    },
    text: {
      fontSize: 12,
      color: "#fff",
      textAlign: "center"
    }
  }
};

Badge.propTypes = {
  color: PropTypes.string,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  textStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  theme: PropTypes.object
};

export default withTheme("Badge", Badge);
