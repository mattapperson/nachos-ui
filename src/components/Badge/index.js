import React from "react";
import PropTypes from "prop-types";
import { View } from "react-native";
import Text from "../typography/Text";
import { withTheme } from "../Theme";

const Badge = props => {
  const { color, style, textStyle, value, theme } = props;

  return (
    <View style={theme.container}>
      <View
        style={[theme.base, color ? { backgroundColor: color } : {}, style]}
      >
        <Text style={[theme.text, textStyle]}>{value}</Text>
      </View>
    </View>
  );
};

Badge.themeConfig = {
  style: {
    container: {
      flexDirection: "row"
    },
    base: {
      padding: 10,
      paddingVertical: 4,
      backgroundColor: "@accentColor",
      borderRadius: 20,
      minWidth: 40
    },
    text: {
      paddingVertical: 0,
      fontSize: 12,
      lineHeight: 16,
      color: "#fff",
      textAlign: "center"
    }
  }
};

Badge.propTypes = {
  color: PropTypes.string,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  textStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  theme: PropTypes.object
};

export default withTheme("Badge", Badge);
