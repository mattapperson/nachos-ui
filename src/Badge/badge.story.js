import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";

import React from "react";
import { View } from "react-native";
import Badge from "./index";

const BadgeExample = () => {
  return <Badge value={123} />;
};

storiesOf("Badge", module).add("basic", withInfo()(BadgeExample));

export default BadgeExample;
