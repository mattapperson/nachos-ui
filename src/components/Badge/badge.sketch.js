import React from "react";
import Badge from "./index";

export const Component = () => {
  return <Badge value={123} />;
};

export const name = "Badge";
