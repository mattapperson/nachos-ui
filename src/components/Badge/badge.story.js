import { storiesOf } from "../../storybook";
import { withInfo } from "../../storybook";

import React from "react";
import Badge from "./index";

const BadgeExample = () => {
  return <Badge value={123} />;
};

storiesOf("Badge", module).add("basic", withInfo()(BadgeExample));

export default BadgeExample;
