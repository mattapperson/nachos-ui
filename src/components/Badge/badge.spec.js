import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import Badge from "../Badge";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Badge value={4} color="red" />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
