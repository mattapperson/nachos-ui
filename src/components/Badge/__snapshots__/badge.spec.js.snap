// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
>
  <div
    className="rn-alignItems-1oszu61 rn-backgroundColor-1g6456j rn-borderTopLeftRadius-f1w8kp rn-borderTopRightRadius-cqzzvf rn-borderBottomRightRadius-1hd4d43 rn-borderBottomLeftRadius-u0ci3n rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-vkv6oe rn-paddingTop-1h8ys4a rn-paddingRight-1qfoi16 rn-paddingBottom-iphfwy rn-paddingLeft-1hfyk0a rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <div
      className="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-1471scf rn-fontFamily-baq3cm rn-fontStyle-13uqrnb rn-fontVariant-ebii48 rn-fontWeight-aruset rn-height-1pexk7n rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
      style={
        Object {
          "color": "rgba(255,255,255,1.00)",
          "fontSize": "12px",
          "lineHeight": "16px",
          "textAlign": "center",
        }
      }
    >
      4
    </div>
  </div>
</div>
`;
